kyo-core/shared/src/main/scala/kyo/System.scala

```diff
@@ -0,0 +1,140 @@
+            reduce(local.use(_.property[E, A](name)))
+
+        def apply[E](name: String, default: => A)(
+            using
+            parser: Parser[E, A],
+            frame: Frame,
+            reduce: Reducible[Abort[E]]
+        ): A < (reduce.SReduced & IO) =
+            reduce(local.use(_.property[E, A](name).map(_.getOrElse(default))))
+    end PropertyOps
+
+    def property[A]: PropertyOps[A] = PropertyOps(())
+
+    def lineSeparator(using Frame): String < IO = local.use(_.lineSeparator)
+    def userName(using Frame): String < IO      = local.use(_.userName)
+    def userHome(using Frame): String < IO      = local.use(_.userHome)
+
+    abstract class Parser[E, A]:
+        def apply(s: String): A < Abort[E]
+
+    object Parser:
+        given (using Frame): Parser[Nothing, String]                    = v => v
+        given (using Frame): Parser[NumberFormatException, Int]         = v => Abort.catching(v.toInt)
+        given (using Frame): Parser[NumberFormatException, Long]        = v => Abort.catching(v.toLong)
+        given (using Frame): Parser[NumberFormatException, Float]       = v => Abort.catching(v.toFloat)
+        given (using Frame): Parser[NumberFormatException, Double]      = v => Abort.catching(v.toDouble)
+        given (using Frame): Parser[IllegalArgumentException, Boolean]  = v => Abort.catching(v.toBoolean)
+        given (using Frame): Parser[NumberFormatException, Byte]        = v => Abort.catching(v.toByte)
+        given (using Frame): Parser[NumberFormatException, Short]       = v => Abort.catching(v.toShort)
+        given (using Frame): Parser[Duration.InvalidDuration, Duration] = v => Abort.get(Duration.parse(v))
+
+        given (using Frame): Parser[IllegalArgumentException, java.util.UUID] =
+            v => Abort.catching(java.util.UUID.fromString(v))
+
+        given (using Frame): Parser[java.time.format.DateTimeParseException, java.time.LocalDate] =
+            v => Abort.catching(java.time.LocalDate.parse(v))
+
+        given (using Frame): Parser[java.time.format.DateTimeParseException, java.time.LocalTime] =
+            v => Abort.catching(java.time.LocalTime.parse(v))
+
+        given (using Frame): Parser[java.time.format.DateTimeParseException, java.time.LocalDateTime] =
+            v => Abort.catching(java.time.LocalDateTime.parse(v))
+
+        given (using Frame): Parser[java.net.URISyntaxException, java.net.URI] =
+            v => Abort.catching(new java.net.URI(v))
+
+        given (using Frame): Parser[java.net.MalformedURLException, java.net.URL] =
+            v => Abort.catching(new java.net.URL(v))
+
+        given [E, A](using p: Parser[E, A], frame: Frame): Parser[E, Seq[A]] =
+            v => Kyo.foreach(v.split(",").toIndexedSeq)(s => p(s.trim()))
+
+        given (using Frame): Parser[IllegalArgumentException, Char] =
+            v => if v.length == 1 then v.charAt(0) else Abort.fail(new IllegalArgumentException("String must have exactly one character"))
```

Comment:
`Abort.when(p(v))(e).map(_ => v)` seems like a common pattern. Perhaps we should make a function.

Reply:
I've added a new `Abort.ensuring` method as a more generic version of `Abort.when`: https://github.com/getkyo/kyo/pull/628/commits/62f2bd1b333808f084576394d74da903ad4a49d3

Reply:
I'm not very happy with the name, let me know if you have suggestions!